contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -95,12 +59,88 @@ bool Matcher::match(absl::optional<absl::string_view> input) {
       },
       &matched);
   if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
-    ENVOY_LOG_MISC(error, "unable to scan, error code {}.", err);
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
   }
 
   return matched;
 }
 
+std::string Matcher::replaceAll(absl::string_view value, absl::string_view substitution) const {
+  // Find matched pair.
+  using Found = std::pair<unsigned long long, unsigned long long>;
+  std::vector<Found> founds;
+  hs_scratch_t* scratch_ = tls_->get()->scratch_;
+  hs_error_t err = hs_scan(
+      som_database_, value.data(), value.size(), 0, scratch_,
+      [](unsigned int, unsigned long long from, unsigned long long to, unsigned int,
+         void* context) -> int {
+        std::vector<Found>* founds = static_cast<std::vector<Found>*>(context);
+        founds->push_back({from, to});
+
+        // Continue searching.
+        return 0;
+      },
+      &founds);
+  if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
+    return std::string(value);
+  }
+
+  // Sort founds. Make sure the longest length found in the front will appear first.
+  std::sort(founds.begin(), founds.end(), [](Found a, Found b) {
+    if (a.first == b.first) {
+      return a.second > b.second;
```

Comment:
I feel like this sorting semantics could be broken out into its own function and unit-tested in isolation from the regex logic. WDYT?

Reply:
Unit tests have covered these codes. But I am willing to extract them into a class member function.

Reply:
Also, I suggest making bounds a struct, so we can implement `operator<` directly for the struct.